### PR TITLE
feat(developer):  double markers once again 🙀

### PR DESCRIPTION
**Pull request by srl295**
- ts side - see also #10563 
- use an 'end' marker for each NFD

#10516 

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/types/src/ldml-keyboard/pattern-parser.ts [hunk 1]
```diff
@@ -275,48 +275,45 @@ export class MarkerParser {
...
-      // 1. we write any markers which match the output string in order,
-      // and pop them off the list as dones
-      while(map2.length > 0 && map2[map2.length-1].ch === p) {
-        const { marker } = map2.pop();
-        if (marker !== constants.marker_no_index) { // if not already written
-          out = MarkerParser.prepend_marker(out, marker, forMatch);
-          written_markers++;
-          // no need to update .marker here, we're about to pop it
-        }
-      }
```
Comment (srl295) [inline common/web/types/src/ldml-keyboard/pattern-parser.ts:305]: I took out this 'quick check' part.  Instead, we attempt to pop off (shorten) the list where possible.

File: common/web/types/test/ldml-keyboard/test-pattern-parser.ts [hunk 2]
```diff
@@ -379,6 +379,12 @@ describe('Test of nfd_markers()', () => {
       ["\u0995\uFFFF\u0008\u0001\u09CB", "\u0995\uFFFF\u0008\u0001\u09C7\u09BE"],
       ["\u0995\u09BE\uFFFF\u0008\u0001\u09C7"],
       ["\u03B5\uFFFF\u0008\u0001\u0344", "\u03B5\uFFFF\u0008\u0001\u0308\u0301"],
+
+      // double marker -
+      ["e\uffff\u0008\u0001\u0300\u0320\u0300", "e\u0320\uffff\u0008\u0001\u0300\u0300"],
+      // double marker - no change
+      ["e\u0320\uffff\u0008\u0001\u0300\u0300", "e\u0320\uffff\u0008\u0001\u0300\u0300"],
+
```
Comment (srl295) [inline common/web/types/test/ldml-keyboard/test-pattern-parser.ts:387]: @jahorton this is your example..

File: common/web/types/src/ldml-keyboard/pattern-parser.ts [hunk 3]
```diff
@@ -340,13 +337,24 @@ export class MarkerParser {
...
         const marker = last_markers[0];
         last_markers = last_markers.slice(1); // pop from front
-        map.push({ ch, marker });
+        map.push({ ch: glue, marker });
+      }
+      // now, push the rest of the glue chars as an NFD sequence.
+      // For example, `\m{m}\u0344` will create the following stream:
+      //  { ch: 0308, end: true}
+      //  { ch: 0308, marker: 1}
+      //  { ch: 0301, end: true}  // added because of decomp
```
Comment (mcdurdin) [inline common/web/types/src/ldml-keyboard/pattern-parser.ts:355]: I see so the marker gets glued to the last character in the decomposition. Definitely slightly unintuitive -- more so than other gluings.

**Review by mcdurdin — APPROVED** [review]
LGTM. My head is hurting now but it seems to all make sense

File: common/web/types/src/ldml-keyboard/pattern-parser.ts (hunk 3 above)
Comment (srl295) [reply common/web/types/src/ldml-keyboard/pattern-parser.ts:355]: Should be FIRST char

File: common/web/types/test/ldml-keyboard/test-pattern-parser.ts (hunk 2 above)
Comment (jahorton) [reply common/web/types/test/ldml-keyboard/test-pattern-parser.ts:387]: So... the marker is `\u0001`?  That's the part before the first entry of the doubled `\u0300`.  (I'd have expected `\uffff`, but that wouldn't make it "my example".)

Assuming so...

```suggestion

      // double marker -
      ["e\uffff\u0008\u0001\u0300\u0320\u0300", "e\u0320\uffff\u0008\u0001\u0300\u0300"],
      // double marker - no change
      ["e\u0320\uffff\u0008\u0001\u0300\u0300", "e\u0320\uffff\u0008\u0001\u0300\u0300"],
      // double marker - in front of second
      ["e\u0320\uffff\u0008\u0300\u0001\u0300", "e\u0320\uffff\u0008\u0300\u0001\u0300"],
```

We want to be sure it's handled right, whichever of the duplicated entries it was originally in front of.

File: common/web/types/src/ldml-keyboard/pattern-parser.ts (hunk 3 above)
Comment (mcdurdin) [reply common/web/types/src/ldml-keyboard/pattern-parser.ts:355]: I'm not entirely following?

`\m{m}\u0344` ==> `\m{m}\u0308\u0301` or `\u0308\m{m}\u0301`?

File: common/web/types/test/ldml-keyboard/test-pattern-parser.ts (hunk 2 above)
Comment (mcdurdin) [reply common/web/types/test/ldml-keyboard/test-pattern-parser.ts:387]: Marker is `\uffff\u0008{marker_id}`.

Comment (jahorton) [reply common/web/types/test/ldml-keyboard/test-pattern-parser.ts:387]: Also, for extra robustness... how well does the solution hold up if the same mark gets repeated 3+ times within the same segmentation cluster?  (Or... is that too niche for us to care about here?)

The case I'd most expect to fail would be something like this - with the marker prepended to the _middle_ instance of the triplicated char.
```typescript
// triple diacritic + marker - reordering needed
["e\u0300\uffff\u0008\u0001\u0300\u0320\u0300", "e\u0320\u0300\uffff\u0008\u0001\u0300\u0300"],
```

File: common/web/types/src/ldml-keyboard/pattern-parser.ts (hunk 3 above)
Comment (srl295) [reply common/web/types/src/ldml-keyboard/pattern-parser.ts:355]: `\m{m}\u0308\u0301` - adding a test.  markers are re-added in reverse order.

File: common/web/types/test/ldml-keyboard/test-pattern-parser.ts (hunk 2 above)
Comment (srl295) [reply common/web/types/test/ldml-keyboard/test-pattern-parser.ts:387]: @jahorton it should hold up - it handles 1:1 all input nfd chars.

Comment (jahorton) [reply common/web/types/test/ldml-keyboard/test-pattern-parser.ts:399]: Then it should hold up with the suggested extra unit test cases, right?

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.259-alpha](https://keyman.com/downloads/releases/alpha/17.0.259)

**Pull request merged**